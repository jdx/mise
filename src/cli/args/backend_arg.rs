use heck::ToKebabCase;
use std::collections::HashMap;
use std::fmt::{Debug, Display};
use std::hash::Hash;
use std::path::PathBuf;

use once_cell::sync::Lazy;

use crate::backend::{unalias_backend, BackendType};
use crate::dirs;
use crate::registry::REGISTRY;

#[derive(Clone, PartialOrd, Ord)]
pub struct BackendArg {
    /// user-specified identifier, "node", "npm:prettier", "cargo:eza", "vfox:version-fox/vfox-nodejs"
    /// multiple ids may point to a single tool, e.g.: "node", "core:node" or "vfox:version-fox/vfox-nodejs"
    /// and "vfox:https://github.com/version-fox/vfox-nodejs"
    pub id: String,
    /// the name of the tool within the backend, e.g.: "node", "prettier", "eza", "vfox-nodejs"
    pub name: String,
    /// type of backend, "asdf", "cargo", "core", "npm", "vfox"
    pub backend_type: BackendType,
    /// ~/.local/share/mise/cache/<THIS>
    pub cache_path: PathBuf,
    /// ~/.local/share/mise/installs/<THIS>
    pub installs_path: PathBuf,
    /// ~/.local/share/mise/downloads/<THIS>
    pub downloads_path: PathBuf,
}

impl<A: AsRef<str>> From<A> for BackendArg {
    fn from(s: A) -> Self {
        let s = s.as_ref();
        if let Some(fa) = FORGE_MAP.get(s) {
            return fa.clone();
        }
        if let Some((backend_type, name)) = s.split_once(':') {
            if let Ok(backend_type) = backend_type.parse() {
                return Self::new(backend_type, name);
            }
        }
        Self::new(BackendType::Asdf, s)
    }
}

impl BackendArg {
    pub fn new(backend_type: BackendType, name: &str) -> Self {
        let name = unalias_backend(name).to_string();
        let id = match backend_type {
            BackendType::Asdf | BackendType::Core => name.clone(),
            backend_type => format!("{backend_type}:{name}"),
        };
        let pathname = id.to_kebab_case();
        Self {
            name,
            backend_type,
            id,
            cache_path: dirs::CACHE.join(&pathname),
            installs_path: dirs::INSTALLS.join(&pathname),
            downloads_path: dirs::DOWNLOADS.join(&pathname),
        }
    }
}

impl Display for BackendArg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.id)
    }
}

impl Debug for BackendArg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, r#"BackendArg("{}")"#, self.id)
    }
}

impl PartialEq for BackendArg {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

impl Eq for BackendArg {}

impl Hash for BackendArg {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}

static FORGE_MAP: Lazy<HashMap<&'static str, BackendArg>> = Lazy::new(|| {
    REGISTRY
        .iter()
        .map(|(short, full)| {
            let (backend_type, name) = full.split_once(':').unwrap();
            let backend_type = backend_type.parse().unwrap();
            let fa = BackendArg::new(backend_type, name);
            (*short, fa)
        })
        .collect()
});

#[cfg(test)]
mod tests {
    use pretty_assertions::{assert_eq, assert_str_eq};

    use super::*;

    #[test]
    fn test_backend_arg() {
        let t = |s: &str, id, name, t| {
            let fa: BackendArg = s.into();
            assert_str_eq!(fa.id, id);
            assert_str_eq!(fa.name, name);
            assert_eq!(fa.backend_type, t);
        };
        let asdf = |s, id, name| t(s, id, name, BackendType::Asdf);
        let cargo = |s, id, name| t(s, id, name, BackendType::Cargo);
        // let core = |s, id, name| t(s, id, name, BackendType::Core);
        let npm = |s, id, name| t(s, id, name, BackendType::Npm);

        asdf("asdf:poetry", "poetry", "poetry");
        asdf("poetry", "poetry", "poetry");
        asdf("", "", "");
        cargo("cargo:eza", "cargo:eza", "eza");
        // core("node", "node", "node");
        npm("npm:@antfu/ni", "npm:@antfu/ni", "@antfu/ni");
        npm("npm:prettier", "npm:prettier", "prettier");
    }

    #[test]
    fn test_backend_arg_pathname() {
        let t = |s: &str, expected| {
            let fa: BackendArg = s.into();
            let actual = fa.installs_path.to_string_lossy();
            let expected = dirs::INSTALLS.join(expected);
            assert_str_eq!(actual, expected.to_string_lossy());
        };
        t("asdf:node", "node");
        t("node", "node");
        t("", "");
        t("cargo:eza", "cargo-eza");
        t("npm:@antfu/ni", "npm-antfu-ni");
        t("npm:prettier", "npm-prettier");
    }
}
