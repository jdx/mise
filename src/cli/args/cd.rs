use clap::{Arg, ArgAction};

#[derive(Clone)]
pub struct Cd;

impl Cd {
    pub fn arg() -> Arg {
        Arg::new("cd")
            .short('C')
            .long("cd")
            .help("Change directory before running command")
            .global(true)
            .action(ArgAction::Set)
            .value_hint(clap::ValueHint::DirPath)
            .value_name("DIR")
    }
}
