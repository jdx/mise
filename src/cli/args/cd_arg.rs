use clap::{Arg, ArgAction};
use std::path::PathBuf;

#[derive(Clone)]
pub struct CdArg;

impl CdArg {
    pub fn arg() -> Arg {
        Arg::new("cd")
            .value_parser(clap::value_parser!(PathBuf))
            .short('C')
            .long("cd")
            .help("Change directory before running command")
            .global(true)
            .action(ArgAction::Set)
            .value_hint(clap::ValueHint::DirPath)
            .value_name("DIR")
    }
}
