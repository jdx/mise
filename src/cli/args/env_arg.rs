use clap::{Arg, ArgAction};
use std::sync::LazyLock as Lazy;

#[derive(Clone, Debug)]
pub struct ProfileArg;

pub static PROFILE_ARG: Lazy<Arg> = Lazy::new(ProfileArg::arg);
pub static ENV_ARG: Lazy<Arg> =
    Lazy::new(|| clap::arg!(-E --env "Set the environment for loading mise.<ENV>.toml files."));

impl ProfileArg {
    fn arg() -> Arg {
        Arg::new("profile")
            .short('P')
            .long("profile")
            .help("Set the profile (environment)")
            .value_parser(clap::value_parser!(String))
            .value_name("PROFILE")
            .action(ArgAction::Set)
            .global(true)
    }
}
