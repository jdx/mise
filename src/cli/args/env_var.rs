use clap::{
    error::{ContextKind, ContextValue, ErrorKind},
    Arg, Command, Error,
};
use std::ffi::OsStr;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct EnvVarArg {
    pub key: String,
    pub value: String,
}

impl EnvVarArg {
    pub fn parse(input: &str) -> Option<Self> {
        input.split_once('=').map(|(k, v)| Self {
            key: k.to_string(),
            value: v.to_string(),
        })
    }
}

#[derive(Debug, Clone)]
pub struct EnvVarArgParser;

impl clap::builder::TypedValueParser for EnvVarArgParser {
    type Value = EnvVarArg;

    fn parse_ref(
        &self,
        cmd: &Command,
        arg: Option<&Arg>,
        value: &OsStr,
    ) -> Result<Self::Value, Error> {
        if let Some(parsed) = EnvVarArg::parse(&value.to_string_lossy()) {
            return Ok(parsed);
        }

        let mut err = clap::Error::new(ErrorKind::ValueValidation).with_cmd(cmd);
        if let Some(arg) = arg {
            err.insert(
                ContextKind::InvalidArg,
                ContextValue::String(arg.to_string()),
            );
        }
        err.insert(
            ContextKind::InvalidValue,
            ContextValue::String(value.to_string_lossy().into()),
        );
        Err(err)
    }
}

#[cfg(test)]
mod tests {
    use super::EnvVarArg;

    #[test]
    fn invalid_value() {
        let res = EnvVarArg::parse("NO_EQUAL_SIGN");
        assert!(res.is_none());
    }

    #[test]
    fn valid_values() {
        let values = [
            ("FOO=", new_arg("FOO", "")),
            ("FOO=bar", new_arg("FOO", "bar")),
        ];

        for (input, want) in values {
            let got = EnvVarArg::parse(input);
            assert_eq!(got, Some(want));
        }
    }

    fn new_arg(key: &str, value: &str) -> EnvVarArg {
        EnvVarArg {
            key: key.to_string(),
            value: value.to_string(),
        }
    }
}
