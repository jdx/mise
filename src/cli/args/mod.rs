pub use cd_arg::CdArg;
pub use env_var_arg::EnvVarArg;
pub use forge_arg::ForgeArg;
pub use log_level_arg::{DebugArg, LogLevelArg, TraceArg};
pub use quiet_arg::QuietArg;
pub use tool_arg::ToolArg;
pub use tool_arg::ToolVersionType;
pub use verbose_arg::VerboseArg;
pub use yes_arg::YesArg;

mod cd_arg;
mod env_var_arg;
mod forge_arg;
mod log_level_arg;
mod quiet_arg;
mod tool_arg;
mod verbose_arg;
mod yes_arg;
