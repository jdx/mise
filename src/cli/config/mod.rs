use clap::Subcommand;
use eyre::Result;

mod generate;
mod ls;

/// [experimental] Manage config files
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "cfg")]
pub struct Config {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// Do not print table header
    #[clap(long, alias = "no-headers", verbatim_doc_comment)]
    no_header: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Ls(ls::ConfigLs),
    Generate(generate::ConfigGenerate),
}

impl Commands {
    pub fn run(self) -> Result<()> {
        match self {
            Self::Ls(cmd) => cmd.run(),
            Self::Generate(cmd) => cmd.run(),
        }
    }
}

impl Config {
    pub fn run(self) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::ConfigLs {
            no_header: self.no_header,
        }));

        cmd.run()
    }
}
