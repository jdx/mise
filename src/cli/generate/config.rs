use crate::Result;
use crate::cli::config::generate;

/// [experimental] Generate a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = generate::AFTER_LONG_HELP)]
pub struct Config {
    #[clap(flatten)]
    generate: generate::ConfigGenerate,
}

impl Config {
    pub async fn run(self) -> Result<()> {
        self.generate.run()
    }
}
