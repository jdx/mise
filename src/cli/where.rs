use eyre::Result;

use crate::cli::args::ToolArg;
use crate::config::Config;
use crate::errors::Error::VersionNotInstalled;
use crate::forge;
use crate::toolset::ToolsetBuilder;

/// Display the installation path for a runtime
///
/// Must be installed.
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Where {
    /// Tool(s) to look up
    /// e.g.: ruby@3
    /// if "@<PREFIX>" is specified, it will show the latest installed version
    /// that matches the prefix
    /// otherwise, it will show the current, active installed version
    #[clap(required = true, value_name = "TOOL@VERSION", verbatim_doc_comment)]
    tool: ToolArg,

    /// the version prefix to use when querying the latest version
    /// same as the first argument after the "@"
    /// used for asdf compatibility
    #[clap(hide = true, verbatim_doc_comment)]
    asdf_version: Option<String>,
}

impl Where {
    pub fn run(self) -> Result<()> {
        let config = Config::try_get()?;
        let runtime = match self.tool.tvr {
            None => match self.asdf_version {
                Some(version) => self.tool.with_version(&version),
                None => {
                    let ts = ToolsetBuilder::new()
                        .with_args(&[self.tool.clone()])
                        .build(&config)?;
                    let v = ts
                        .versions
                        .get(&self.tool.forge)
                        .and_then(|v| v.requests.first())
                        .map(|r| r.version());
                    self.tool.with_version(&v.unwrap_or(String::from("latest")))
                }
            },
            _ => self.tool,
        };

        let plugin = forge::get(&runtime.forge);

        match runtime
            .tvr
            .as_ref()
            .map(|tvr| tvr.resolve(plugin.as_ref(), false))
        {
            Some(Ok(tv)) if plugin.is_version_installed(&tv) => {
                miseprintln!("{}", tv.install_path().to_string_lossy());
                Ok(())
            }
            _ => Err(VersionNotInstalled(
                runtime.forge.to_string(),
                runtime.tvr.map(|tvr| tvr.version()).unwrap_or_default(),
            ))?,
        }
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
    # Show the latest installed version of node
    # If it is is not installed, errors
    $ <bold>mise where node@20</bold>
    /home/jdx/.local/share/mise/installs/node/20.0.0

    # Show the current, active install directory of node
    # Errors if node is not referenced in any .tool-version file
    $ <bold>mise where node</bold>
    /home/jdx/.local/share/mise/installs/node/20.0.0
"#
);

#[cfg(test)]
mod tests {
    use crate::dirs;
    use crate::test::reset;
    use test_log::test;

    #[test]
    fn test_where() {
        reset();
        assert_cli!("install");
        let stdout = assert_cli!("where", "tiny");
        assert_str_eq!(
            stdout.trim(),
            dirs::DATA.join("installs/tiny/3.1.0").to_string_lossy()
        );
    }

    #[test]
    fn test_where_asdf_style() {
        reset();
        assert_cli!("install", "tiny@2", "tiny@3");
        assert_cli_snapshot!("where", "tiny", "2");
        assert_cli_snapshot!("where", "tiny", "3");
    }

    #[test]
    fn test_where_alias() {
        reset();
        assert_cli!("install", "tiny@my/alias");
        let stdout = assert_cli!("where", "tiny@my/alias");
        assert_str_eq!(
            stdout.trim(),
            dirs::DATA.join("installs/tiny/3.0.1").to_string_lossy()
        );
        assert_cli!("uninstall", "tiny@my/alias");
    }

    #[test]
    fn test_where_not_found() {
        reset();
        let err = assert_cli_err!("where", "tiny@1111");
        assert_snapshot!(err, @"tiny@1111 not installed");
    }
}
