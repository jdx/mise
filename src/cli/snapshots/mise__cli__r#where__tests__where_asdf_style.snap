---
source: src/cli/where.rs
expression: output
snapshot_kind: text
---
~/data/installs/tiny/2.1.0
